### DISCUSSION TITLE
Impersonate users (Credentials/Google)

### DISCUSSION THREAD

**Discussion by jonasgroendahl** [nextauthjs/next-auth#2947] [Help]
Hi all,

I'm thinking about implementation an impersonation feature where an admin can login as other users. In my app I'm allowing to login with Google and Credentials(username and password).

How to go about implementing such a feature with next-auth? Like for example it wouldn't be possible for an admin to login with someone's else Google account but perhaps there's a way to mutate the session with next-auth? I haven't been able to find it searching the docs.

Anyone with experience implementing this type of functionality?

**Comment by balazsorban44** [comment]
I think you are talking about [Role-Based Access Control](https://en.wikipedia.org/wiki/Role-based_access_control). There are many ways going about that, the easiest is if you add a `role` property on sign-up/sign-in to users.

**Reply by jonasgroendahl** [reply]
Perhaps,

I'm just thinking say I'm logged in as User 1 (Admin) and I want to see exactly what the User 2 sees when logged in to the app. 

So there's a feature inside the app that allow User 1 to **sign()** as User 2 replacing the current session to that of User 2 so all getSession() and useSession() will temporarily return the information of User 2. 

User 1 will of course need an option to return to his login, replacing the session with User 2 back to User 1.

That's the idea :)

**Reply by balazsorban44** [reply]
Yes, that's exactly what RBAC let's you do. People with higher roles can "impersonate" normal users. If you want to mimic their UI, just show an extra button for admins that says "view as user" or something.

**Reply by balazsorban44** [reply]
Actually logging in with someone else's account though wouldn't be much of a privacy friendly solution, and in most cases impossible. You cannot and should not know your users' credentials, so you might want to avoid that.

Role based access can accommodate for your requirements.

**Reply by jonasgroendahl** [reply]
Alright cool :) 

Any tips on implementing this Next-auth? Doesn't seem to be that straight forward. There would need to be a way to mutate the session on the demand right? I can add custom fields such as role as you said to the session, i already added some fields but how do navigate between sessions using next-auth?

**Reply by balazsorban44** [reply]
No, you don't have to mutate the session. upon session creation, add the correct role to the user based on some criteria (like their e-mail), unless your underlying provider can add roles to the user already. Google can't I think, but providers like Keycloak, IDS4, Auth0 for example can, it all depends. Check out the docs for the jwt and session callbacks:

https://next-auth.js.org/configuration/callbacks#jwt-callback
https://next-auth.js.org/configuration/callbacks#session-callback

This is a very vague question, so I don't have the capacity to give you a full solution, but I would say it is very straightforward with `next-auth`. :smile:

https://simplernerd.com/next-auth-global-session this might be useful

**Reply by jonasgroendahl** [reply]
Thanks I will have a look at the links!

**Reply by rnbrady** [reply]
User impersonation is an established concept which is distinct from but dependent on RBAC.

**Reply by ian** [reply]
Any update on how to do this? Links don't provide much aside from overview.

Ideally there would be a way to just tap into the auth system server side with a `signIn(user)` and that'll replace the current session with the associated user. Server side only of course.

**Reply by ian** [reply]
@balazsorban44 Should we move the impersonation feature request to a different discussion? All other requests seem to point here but this ticket doesn't actually talk about impersonation, it talks about role assignments upon signin/register, which is very different.

Unless I'm mistaken, the original ticket was asking for implementation details on impersonating a user, not role assignment. For comparison, you can do this in:
- clerk.dev https://clerk.com/docs/authentication/user-impersonation
- blitzjs https://blitzjs.com/docs/impersonation

**Reply by Untel** [reply]
Well, if you use a DB adapter you can do (but will not trigger session, signin... callbacks as the user would)
```ts
prisma.session.updateMany({
     where: { userId: ctx.session.user.id }, // current user -> you
     data: { userId: userIdToImpersonate }, // user you want to control
 });
```

**Reply by mistersingh179** [reply]
fyi - this will update all sessions of current user to the new user. ideally we may want to update only that one session. this way in different browsers we can still have the original user as-is

**Reply by Untel** [reply]
Yes but sadly ctx.session.id does not exist (in the typescript declaration)

**Comment by zocoi** [comment]
> Well, if you use a DB adapter you can do (but will not trigger session, signin... callbacks as the user would)
> 
> ```ts
> prisma.session.updateMany({
>      where: { userId: ctx.session.user.id }, // current user -> you
>      data: { userId: userIdToImpersonate }, // user you want to control
>  });
> ```

any idea on how to do it with JWT token session which is the default session store?

**Reply by eznix86** [reply]
Since a user is represented by

```json
{
  user: {
    name: string
    email: string
    image: string
  },
  expires: Date // This is the expiry of the session, not any of the tokens within the session
}

```
Maybe we can have an extra field like: `impersonator: true` and other fields also.

then we can trigger an update see [](https://next-auth.js.org/getting-started/client#updating-the-session) to update the session, what do you think ?

**Comment by Coderamrin** [comment]
I implemented it with the Credentials provider. 
you can have multiple credentials provider with different ids
https://next-auth.js.org/configuration/providers/credentials

```javascript
CredentialsProvider({
			name: "impersonate",
			id: "impersonate",
			credentials: {
				adminEmail: {
					label: "Admin Email",
					type: "text",
					placeholder: "Jhondoe@gmail.com",
				},
				userEmail: {
					label: "User Email",
					type: "text",
					placeholder: "Jhondoe@gmail.com",
				},
			},

			async authorize(credentials) {
				if (!credentials?.adminEmail || !credentials?.userEmail) {
					throw new Error("User email or Admin email is missing");
				}

				const admin = await prisma.user.findUnique({
					where: {
						email: credentials.adminEmail.toLocaleLowerCase(),
					},
				});

				const user = await prisma.user.findUnique({
					where: {
						email: credentials.userEmail.toLocaleLowerCase(),
					},
				});

				if (!admin || admin.role !== "ADMIN") {
					throw new Error("Access denied");
				}

				// if user was not found
				if (!user) {
					throw new Error("No user found");
				}
				return user;
			},
		}),
```



call the `signIn()` function with the credentials id

```

signIn("impersonate", {
			adminEmail: session?.user?.email, //the admin email
			userEmail: user?.email, //the user you want to impersonate
		}).then((callback) => {
			if (callback?.error) {
				console.log(callback?.error);
				toast.error(callback.error);
			}

			if (callback?.ok && !callback?.error) {
				toast.success("Logged in successfully");
				router.refresh();
			}
		});

```

**Reply by yagudaev** [reply]
Can't `signIn` be called from the client side as well? If so, wouldn't that mean any users can become another user as long as they know the admin's email?

**Reply by lswartsenburg** [reply]
Yessir! This is not a safe solution

**Reply by mistersingh179** [reply]
inside the action, we can check for which user is logged in and making that call. perhaps you only allow certain users the ability to impersonate.

**Reply by yagudaev** [reply]
A simple way to implement this feature is to use the auth token given to you in a passwordless login/magic link -- the token typically emailed to the user.

You can either send it to yourself as the admin for a particular user or show the link in your ui and click on it to be logged in as that user.

I've ended up doing something more complicated, with modifying the session. Only realized this after my friend told me about how he did it.

**Reply by ChrisB1123** [reply]
Just to expand on @yagudaev's suggestion because I got lost on a couple of different versions of the docs

I followed this to get the magic link adapter set up. I was on next-auth v5.0.0-beta.4, and needed to upgrade to 5.0.0-beta.22 to have the correct providers available.
https://authjs.dev/getting-started/authentication/email

Then you customise how the email is sent using these steps
https://authjs.dev/getting-started/providers/resend#customization

And to redirect the magic link to the user who's currently logged in I just added
```
  const session = await auth()
  const user = session?.user
```

at the top of `sendVerificationRequest()`.

And then finally edit the email api call so you have

```
    body: JSON.stringify({
      from: provider.from,
      to: user.email,                     <<< Edit this line
      subject: `Sign in to ${host}`,
      html: html({ url, host }),
      text: text({ url, host }),
    }),
```

You also probably want to add a role to the user and check they're an admin before actually sending the email though

**Reply by Geczy** [reply]
here's how i had to do it with plain next auth, to be able to securely get the current logged in user's ID. i'm going to update the prisma query to confirm this role can impersonate

```ts
const extractCookieValue = (cookieHeader: string | string[], name: string) => {
  const cookieStringFull = Array.isArray(cookieHeader)
    ? cookieHeader.find((header) => header.includes(name))
    : cookieHeader
  return name + cookieStringFull?.split(name)[1].split(';')[0]
}


export const authOptions: NextAuthOptions = {
  adapter: PrismaAdapter(prisma),
  session: {
    strategy: 'jwt',
  },
  pages: {
    signIn: '/login',
  },
  providers: [
    CredentialsProvider({
      name: 'impersonate',
      id: 'impersonate',
      credentials: {
        channelToImpersonate: {
          label: 'Channel ID',
          type: 'text',
          placeholder: '1234',
        },
      },

      async authorize(credentials, req) {
        if (
          !credentials?.channelToImpersonate ||
          !Number.parseInt(credentials.channelToImpersonate)
        ) {
          throw new Error('No channel ID provided')
        }

        const headerCookies = req.headers?.cookie
        let currentLoggedInUserId: string | undefined

        try {
          const secureCookie =
            process.env.NEXTAUTH_URL?.startsWith('https://') ??
            !!process.env.VERCEL
          const cookieName = secureCookie
            ? '__Secure-next-auth.session-token'
            : 'next-auth.session-token'
          const sessionToken = extractCookieValue(headerCookies, cookieName)
          const actualToken = await decode({
            secret: process.env.NEXTAUTH_SECRET,
            token: sessionToken.replace(`${cookieName}=`, ''),
          })
          currentLoggedInUserId = actualToken.id
        } catch (e) {
          console.error(e)
          captureException(e)
        }

        if (!currentLoggedInUserId) {
          throw new Error('Access denied')
        }

        const data = await prisma.account.findUnique({
          select: {
            user: true,
          },
          where: {
            providerAccountId: credentials.channelToImpersonate,
          },
        })

        if (!data) {
          throw new Error('Access denied')
        }

        return { ...data?.user, currentLoggedInUserId }
      },
    }),
```

**Comment by musjj** [comment]
With Lucia, you can easily do it like this:

```ts
const session = await lucia.createSession(userId, {});
const sessionCookie = lucia.createSessionCookie(session.id);
cookies().set(
  sessionCookie.name,
  sessionCookie.value,
  sessionCookie.attributes,
);
```

Is there an equivalent API for Auth.js?

**Reply by yagudaev** [reply]
From a quick look at Lucia, it looks way better than auth.js. Auth.js is probably the worst auth library I have ever used. Why not just stick with Lucia?

I am planning to get rid of Auth.js and implement my own auth. It isn't that hard.

**Comment by misterbit-pro** [comment]
Hello,

is ther any update to implementing impersonation in next auth? Because the OAuth 2.0 standard supports token exhange https://cloudentity.com/developers/basics/oauth-extensions/token-exchange/#how-token-exchange-works wich keycloak f.e. also implements as its a safe way to do impersonation, this would be great if it cloud be added to next-auth as a custom implementation over a credentials provider is a bit tricky as the User object wich gets returned in the authorize method is very limited and has no support to save the exchanged token

**Comment by estani** [comment]
for v5.beta.25 (not sure since when it works) using jwt session (documentation is still terrible... but it works and makes sense, aka might be stable):

Client part to send the info to the backen (to the jwt callback)
```typescript
'use client'
import { SessionProvider, useSession } from "next-auth/react"
import { useRouter } from "next/navigation"

function WrappedButtonImpersonate({ userId }: { userId: string | null }) {
  const session = useSession();
  const router = useRouter();
  async function onClick() {
    await session.update({ impersonate: userId });
    router.refresh();
  }
  return <button  onClick={onClick}>{userId === null && 'Stop '}Impersonate</button>
}

export function ButtonImpersonate({ userId }: { userId: string }) {
  return (
    <SessionProvider>
      <WrappedButtonImpersonate userId={userId} />
    </SessionProvider>
  )
}

export function ButtonStopImpersonation() {
  return (
    <SessionProvider>
      <WrappedButtonImpersonate userId={null} />
    </SessionProvider>
  )
}
```
jwt callback to check, handle and update session (working excerpt)
```typescript
 jwt({ token, user, trigger, session }) {
  switch (trigger) {
    case 'update':
      // validate what you need (probably token.isAdmin or something)
      if (session.impersonate) {
        // impersonate (if you need to load things, make this callback async, it works)
        token.userId = session.impersonate;
      } else {
        // stop impersonating;
        token.userId = token.realUserId;
      }
      break;
    case 'signIn':
      token.userId = user.id;
      token.realUserId ||= user.id;  // Keep track of the original user
      break
  }
  return token
},
session({ session, token }) {
  // Enhance session with user details
  session.user.id = token.userId;
}
```
